src/main/flight/imu.c

```diff
@@ -261,150 +243,158 @@ static void imuCheckAndResetOrientationQuaternion(const float ax, const float ay
             // magnetometer error is cross product between estimated magnetic north and measured magnetic north (calculated in EF)
-            const float ez_ef = -(hy * bx);
+            vectorCrossProduct(&vErr, &vMag, &vNorth);
 
-            // Rotate mag error vector back to BF and accumulate
-            ex = rMat[2][0] * ez_ef;
-            ey = rMat[2][1] * ez_ef;
-            ez = rMat[2][2] * ez_ef;
+            // Rotate error back into body frame
+            rotateVectorByQuaternion(&vErr, &vErr, &orientation);
         }
         else if (useCOG) {
+            t_fp_vector vHeadingEF;
+
             // Use raw heading error (from GPS or whatever else)
             while (courseOverGround >  M_PIf) courseOverGround -= (2.0f * M_PIf);
             while (courseOverGround < -M_PIf) courseOverGround += (2.0f * M_PIf);
 
             // William Premerlani and Paul Bizard, Direction Cosine Matrix IMU - Eqn. 22-23
             // (Rxx; Ryx) - measured (estimated) heading vector (EF)
-            // (cos(COG), sin(COG)) - reference heading vector (EF)
+            // (-cos(COG), sin(COG)) - reference heading vector (EF)
+
+            // Compute heading vector in EF from scalar CoG
+            t_fp_vector vCoG = { .V = { -cos_approx(courseOverGround), sin_approx(courseOverGround), 0.0f } };
+
+            // Rotate Forward vector from BF to EF - will yield Heading vector in Earth frame
+            rotateVectorByQuaternionInv(&vHeadingEF, &vNorth, &orientation);
+            vHeadingEF.V.Z = 0.0f;
+
+            // Normalize to unit vector
+            vectorNormalize(&vHeadingEF, &vHeadingEF);
+
             // error is cross product between reference heading and estimated heading (calculated in EF)
-            const float ez_ef = - sin_approx(courseOverGround) * rMat[0][0] - cos_approx(courseOverGround) * rMat[1][0];
+            vectorCrossProduct(&vErr, &vCoG, &vHeadingEF);
 
-            ex = rMat[2][0] * ez_ef;
-            ey = rMat[2][1] * ez_ef;
-            ez = rMat[2][2] * ez_ef;
-        }
-        else {
-            ex = 0;
-            ey = 0;
-            ez = 0;
+            // Rotate error back into body frame
+            rotateVectorByQuaternion(&vErr, &vErr, &orientation);
         }
 
         // Compute and apply integral feedback if enabled
         if (imuRuntimeConfig.dcm_ki_mag > 0.0f) {
             // Stop integrating if spinning beyond the certain limit
             if (spin_rate_sq < sq(DEGREES_TO_RADIANS(SPIN_RATE_LIMIT))) {
-                integralMagX += imuRuntimeConfig.dcm_ki_mag * ex * dt;    // integral error scaled by Ki
-                integralMagY += imuRuntimeConfig.dcm_ki_mag * ey * dt;
-                integralMagZ += imuRuntimeConfig.dcm_ki_mag * ez * dt;
+                t_fp_vector vTmp;
+
+                // integral error scaled by Ki
+                vectorScale(&vTmp, &vErr, imuRuntimeConfig.dcm_ki_mag * dt);
+                vectorAdd(&vIntegralMag, &vIntegralMag, &vTmp);
```

Comment:
There should probably be only one integral error accumulator.
Two integrators may drift in opposite direction - total error will be small, but error in each integrator large, possibly causing problem.

Also integrator is here to remove gyro drift. Not sure, but it's gain is more related to gyro properties (maximum drift) than to reference source (and its noise)...